server/container/guice/jpa-common-guice/src/main/java/org/apache/james/modules/data/JPAConfiguration.java

```diff
@@ -18,67 +18,138 @@
  ****************************************************************/
 package org.apache.james.modules.data;
 
-import org.apache.james.backends.jpa.JPAConstants;
+import java.util.Objects;
+import java.util.Optional;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Preconditions;
 
 public class JPAConfiguration {
 
-    public static Builder builder() {
-        return new Builder();
+    public static class Credential {
+        private static final Logger LOGGER = LoggerFactory.getLogger(Credential.class);
+
+        public static Optional<Credential> of(String username, String password) {
+            if (Objects.isNull(username) && Objects.isNull(password)) {
+                LOGGER.debug("username and password are missing, returns no credential by default");
+                return Optional.empty();
+            }
+
+            return Optional.of(new Credential(username, password));
+        }
+
+        private final String username;
+        private final String password;
+
+        private Credential(String username, String password) {
+            Preconditions.checkArgument(Objects.nonNull(username) && Objects.nonNull(password),
+                "username and password for connecting to database have to be specified");
+            this.username = username;
+            this.password = password;
+        }
+
+        public String getUsername() {
+            return username;
+        }
+
+        public String getPassword() {
+            return password;
+        }
     }
 
-    public static class Builder {
-        private String driverName;
-        private String driverURL;
-        private boolean testOnBorrow;
-        private int validationQueryTimeoutSec = JPAConstants.VALIDATION_NO_TIMEOUT;
-        private String validationQuery;
+    @FunctionalInterface
+    public interface RequireDriverName {
+        RequireDriverURL driverName(String driverName);
+    }
 
+    @FunctionalInterface
+    public interface RequireDriverURL {
+        ReadyToBuild driverURL(String driverUrl);
+    }
 
-        public Builder driverName(String driverName) {
-            this.driverName = driverName;
-            return this;
+    public static class ReadyToBuild {
+        private final String driverName;
+        private final String driverURL;
+
+        private Optional<Credential> credential;
+        private Optional<Boolean> testOnBorrow;
+        private Optional<Integer> validationQueryTimeoutSec;
+        private Optional<String> validationQuery;
+
+        private ReadyToBuild(String driverName, String driverURL, Optional<Credential> credential) {
+            this(driverName, driverURL, credential, Optional.empty(), Optional.empty(), Optional.empty());
```

Comment:
where is it used ?

Reply:
I used it in `static builder()` to expose a default builder with empty optional properties

Reply:
no need this constructor, deleted